consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java

```diff
@@ -0,0 +1,202 @@
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), prepare);
+
+    gossiper.injectCommit(roundId, block.getHash());
+    final IbftExtraData extraData = IbftExtraData.decode(block.getHeader().getExtraData());
+    final Hash commitHash =
+        IbftBlockHashing.calculateDataHashForCommittedSeal(block.getHeader(), extraData);
+    final Signature commitSeal = SECP256K1.sign(commitHash, gossiper.getNodeKeyPair());
+    final SignedData<CommitPayload> commit =
+        msgFactory.createSignedCommitPayload(roundId, block.getHash(), commitSeal);
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), commit);
+
+    final SignedData<RoundChangePayload> roundChange =
+        msgFactory.createSignedRoundChangePayload(roundId, Optional.empty());
+    final RoundChangeCertificate roundChangeCert =
+        new RoundChangeCertificate(Collections.singleton(roundChange));
+    final SignedData<NewRoundPayload> newRound =
+        msgFactory.createSignedNewRoundPayload(roundId, roundChangeCert, proposal);
+    gossiper.injectNewRound(roundId, roundChangeCert, proposal);
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), newRound);
+
+    gossiper.injectRoundChange(roundId, Optional.empty());
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), roundChange);
+  }
+
+  @Test
+  public void messageIsOnlyGossipedOnce() {
+    gossiper.injectProposal(roundId, block);
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), proposal);
+
+    gossiper.injectProposal(roundId, block);
+    assertPeersReceivedNoMessages(roles.getNonProposingPeers());
+
+    gossiper.injectProposal(roundId, block);
+    assertPeersReceivedNoMessages(roles.getNonProposingPeers());
+  }
+
+  @Test
+  public void messageWithUnknownValidatorIsNotGossiped() {
+    final KeyPair unknownKeyPair = KeyPair.generate();
+    final MessageFactory unknownMsgFactory = new MessageFactory(unknownKeyPair);
+    final SignedData<ProposalPayload> unknownProposal =
+        unknownMsgFactory.createSignedProposalPayload(roundId, block);
+
+    gossiper.injectMessage(ProposalMessageData.create(unknownProposal));
+    assertPeersReceivedNoMessages(roles.getNonProposingPeers());
+  }
+
+  @Test
+  public void messageIsNotGossipedToSenderOrCreator() {
+    final ValidatorPeer msgCreator = roles.getNonProposingPeer(0);
+    final MessageFactory peerMsgFactory = msgCreator.getMessageFactory();
+    final SignedData<ProposalPayload> proposalFromPeer =
+        peerMsgFactory.createSignedProposalPayload(roundId, block);
+
+    gossiper.injectMessage(ProposalMessageData.create(proposalFromPeer));
+
+    final List<ValidatorPeer> validators = new ArrayList<>(roles.getNonProposingPeers());
+    validators.remove(msgCreator);
+    assertPeersContainsReceivedMessages(validators, proposalFromPeer);
+    assertPeersReceivedNoMessages(ImmutableList.of(roles.getProposer(), msgCreator));
```

Comment:
Id suggest having the two assert cases separately. It'll make life easier to debug from Unit failure messages (i.e. without needing to run the code locally)

Reply:
The other case where testing that sender is not gossiped to is tested in the other tests as in those tests the sender and the creator are the same. We can't avoid testing that case. What this test is doing differently is testing the case where the sender and the message creator are different nodes and that neither are gossiped too.

Reply:
Right, I understand ...my issue is actually with the underlying function.